tools/firmware-utils/src/xorimage.c

```diff
@@ -100,6 +107,17 @@ int main(int argc, char **argv)
 		usage();
 	}
 
+	if (is_hex_pattern) {
+		if (p_len % 2 != 0) {
+			fprintf(stderr, "the number of characters (hex) is incorrect\n");
+			usage();
+		}
+
+		for (int i = 0; i < (p_len / 2); i++) {
+			sscanf(pattern + (i * 2), "%2x", &hex_buf);
+			hex_pattern[i] = (char)hex_buf;
```

Comment:
Because hex_pattern is declared as

      char *hex_pattern = default_pattern; // = "12345678";

the hex_pattern can store at most 8 bytes.

So, this can/will cause a "segmentation fault", if the provided hex pattern is big enough.

    $ ./xorimage -i xorimage -o xorimage-xor -x -p "$(seq 0 8000)"
    Segmentation fault

Reply:
Hmm...sorry, I confused...

As a pure question (due to ignorance), why is it no problem to pass a 32-byte string to ```pattern``` declared similarly by an 8-byte character string?
In that case, when converted from a character string to a hexadecimal value and passed to ```hex_pattern```, eventually it has a 16-byte value...

Reply:
Thanks for your detailed advice, it seems I got a rough understanding...
Then, how long should I limit ```hex_pattern``` to?

Reply:
> Then, how long should I limit hex_pattern to?

There's already kind of a limit. Since the pattern is passed as a parameter to the xorimage program, the pattern has to go through a execve system call which enforces a string length
check of MAX_ARG_STRLEN.

https://elixir.bootlin.com/linux/v4.20-rc7/source/include/uapi/linux/binfmts.h#L15
    /*
     * These are the maximum length and maximum number of strings passed to the
     * execve() system call.  MAX_ARG_STRLEN is essentially random but serves to
     * prevent the kernel from being unduly impacted by misaddressed pointers.
     * MAX_ARG_STRINGS is chosen to fit in a signed 32-bit integer.
     */
    #define MAX_ARG_STRLEN (PAGE_SIZE * 32)

On a 4K PAGE_SIZE System this would be 131072 / 128 KIB.

Reply:
Hmm...sorry, I searched and investigated about ```execve```, but I did not know how to use ```MAX_ARG_STRLEN``` through it from C source...

Reply:
> Hmm...sorry, I searched and investigated about `execve`, but I did not know how to use `MAX_ARG_STRLEN` through it from C source...

Hm, I'm getting the feeling that we are unfortunately talking past each other at this point :sob: .

The pattern string can't exceed MAX_ARG_STRLEN by design. So you could just allocate a memory buffer on the heap which holds (strlen(pattern) / 2) bytes and free it after use. But you could also declare something like char hex_pattern[128] buffer and check that (strlen(pattern) / 2) does not exceed the sizeof(hex_pattern). Or you can come up with another option. As long as it passes the fuzzing I do it's "just fine" :+1: .

Reply:
I see...
Then, how about with the following changes? (Almost as your advice is...)
```char *hex_pattern = default_pattern;``` -> ```char hex_pattern[128];```

and add
```
		if ((p_len / 2) >= sizeof(hex_pattern)) {
			fprintf(stderr, "provided hex pattern is too long\n");
			usage();
		}
```